Proposal for naming convention to unite SDK and BAG

sporrasm opened the pull request:
This commit adds a warning print if the generator is named something other than <generator_name>_gen.

Also, I performed some optimization to adding the generators to path. The old version used os.walk, which goes through EVERY directory in BAG_WORK_DIR and searches for a file __init__.py. This took appoximately 30 seconds, since it went through all of the PEX, LVS, etc. directories and all associated subdirectories as well. If we rely on the fact that the generators are always in fixed location (this should be always true, right?), e.g. BAG_WORK_DIR/gen/gen/__init__.py, then we can add only those paths matching that pattern. Now it happens in a fraction of a second (as it should) and works from both the SDK and BAG.

Summary: this commit "forces" naming convention of the generators so that they can be used in the SyDeKick. Also, optimizations to performance. @mkosunen @vlahtin

Commit:
Proposal for naming convention to unite SDK and BAG

Also, made adding generators to path much more efficient.

sporrasm commented:
Oh, and I will device a script to change the generators to the current naming convention.

Commit:
Fix rename module

sporrasm commented:
It seems that we already have a renaming script. I had to remove one sed operation to get it working (see 
ebf4e54 ), since it was performed twice for some reason. @mkosunen ready for review

mkosunen committed:
Reword warning

mkosunen committed:
Fixup

mkosunen reviewed (approved):
Tested. Works. Can be merged.

mkosunen commented:
This seems to print the warning for generator named ringoscillator_gen. Sometihng is off.

mkosunen commented:
I revoke my approval :) Some more work is needed. The naming conventions should be:

The name of the generator is <generator>_gen, that provides and instance of <generator> (not instance of <generator>_gen).

sporrasm commented:
@mkosunen So, where are we now with this? I'm very confused: you said that generator python packages should end with _gen (or something similar). For me, warnings about ill-named packages are in fact printed correctly. The warning statement is printed before the addition of a particular package and not after, so that cause some confusion as to which package the warning is actually about? We could add the actual generator name to the warning message if that helps.

mkosunen committed:
Naming convention verified

Naming convention supported as follows:

- Generator inverter should be in directory inverter_gen
- Python package and the master class is inverter_gen described in inverter_gen/inverter_gen/__init__.py
- It generates an instance of inverter with name inverter to virtuoso
library inverter_generated

mkosunen commented:
@sporrasm See the commit message 64f28d1 Is it confusing enough :)
I added the package name to the warning.

Shortly: Python package inverter_gen generates inverter, (not inverter_gen), to virtuoso library inverter_generated with name inverter (not inverter_gen). I.e '_gen' is only a suffix of the python generator package, not a suffix of the desing name.

sporrasm commented:
I see what you mean now. So, in essence __init__.py should contain \<name\>_gen class and schematic and layout generator classes should still be named with only \<name\>. This means that the init_bag_module.sh script has to be modified slightly (not a problem), if we want to stick to using that.

mkosunen commented:
I did the modifications already. These sundays  bore me to death :)

mkosunen commented:
Renaming script may need to be modified, so that if you rename to something that ends with '_gen' that is taken into account.
We could actually force this policy by making renaming script to always append the generator package names with '_gen', if the package name does not end with '_gen'.

sporrasm commented:
Ok. I will test it on my side once you push. Were/are the messages printed incorrectly still? Or did the order confuse you?

mkosunen commented:
I think the commit 64f28d1 is in form I intented. tested it with bot old and new naming convention.

mkosunen commented:
Let me know if something is broken. Renaming script supposedly is.

Commit:
Fixes to naming of layout and schematic generators

sporrasm commented:
I made some changes to init_bag_module to reflect this naming scheme. I did test this in a brief manner (so there may bugs), the new naming convention is now enforced if \_gen is missing from the end of the module name. In addition, templates and generated layouts are now called generator_\<purpose\> rather than generator_gen_\<purpose\>. For our example case of ringoscillator, the generated layout would now be in library ringoscillator_generated under cell ringoscillator. Probably related to issue #7.

mkosunen commented:
Can we merge?

sporrasm commented:
Green light for merge from me.
